File: cpp/session/TLSSession.cpp
#include "session/TLSSession.hpp"

#include "se/CLTRID.hpp"
#include "se/PollRequestCommand.hpp"
#include "se/LoginCommand.hpp"
#include "se/LogoutCommand.hpp"
#include "se/ReceiveSE.hpp"
#include "se/Response.hpp"
#include "se/ResultCode.hpp"
#include "common/ErrorPkg.hpp"
#include "session/CertificateUserMismatchException.hpp"
#include "session/UserPassMismatchException.hpp"
#include "session/SessionLimitExceededException.hpp"
#include "session/UninitialisedSessionException.hpp"
#include "session/GreetingException.hpp"
#include "session/TLSContext.hpp"
#include "session/TLSSocket.hpp"
#include "common/Logger.hpp"
#include "session/Timer.hpp"
#include "xml/ParsingException.hpp"
#include "common/AutoMutex.hpp"
#include "session/SessionProperties.hpp"
#include "se/Greeting.hpp"

#include <xalanc/XalanDOM/XalanText.hpp>
#include <sys/time.h>
#include <errno.h>

using namespace std;
using namespace xalanc;

namespace {
    PollRequestCommand& poll()
    {
        static PollRequestCommand poll;
        return poll;
    }

    const string& pollXML()
    {
        static const string str(poll().toXML());
        return str;
    }
}

TLSSession::TLSSession(SessionProperties* props)
    throw (SessionConfigurationException)
    : _inUse(false), _isOpen(false), _isInvalid(true), commandCounter(NULL)
{
    pthread_mutex_init(&mtx, NULL);
    pthread_cond_init(&cond, NULL);
    configure(props);
    
    const string pname("com.ausregistry.cpptoolkit.session");
    debugLogger = Logger::getLogger(pname + ".debug");
    supportLogger = Logger::getLogger(pname + ".support");
    userLogger = Logger::getLogger(pname + ".user");

    parser = auto_ptr<XMLParser>(new XMLParser);
}

TLSSession::~TLSSession()
{
    pthread_mutex_destroy(&mtx);
    if (pthread_cond_destroy(&cond) == EBUSY)
    {
        try
        {
            supportLogger->LOG_SEVERE(
                "Condition variable for TLSSession destroyed while in use.");
        }
        catch (...)
        { }
    }
    // default dtor for TLSSocket is sufficient.
}

void TLSSession::configure(SessionProperties* props)
    throw (SessionConfigurationException)
{
    hostName = props->getHostname();
    port = props->getPort();
    username = props->getClientID();
    password = props->getClientPW();
    eppVersion = props->getVersion();
    language = props->getLanguage();
    objURIs = props->getObjURIs();
    extURIs = props->getExtURIs();
    acquireTimeout = props->getAcquireTimeout();
    commandCounter.reset(new CommandCounter(props->getCommandLimitInterval()));
    validationEnabled = props->enforceStrictValidation();

    try
    {
        ctx = std::auto_ptr<TLSContext>(
                new TLSContext(props->getPrivateKeyFilename(),
                               props->getCertFilename(),
                               props->getCAFilename(),
                               props->getPrivateKeyPassphrase()));
    }
    catch (const EPPException& e)
    {
        SessionConfigurationException sce("Error during TLSSession::configure.");
        sce.causedBy(e);
        throw sce;
    }
}

void TLSSession::open() throw (SessionOpenException)
{
    debugLogger->LOG_FINEST("enter");
    _isInvalid = true;
    try
    {
        openSocket();
        processGreeting();
        login();
        _isOpen = true;
        _isInvalid = false;
    }
    catch (EPPException& e)
    {
        userLogger->LOG_WARNING(e.getMessage());
        SessionOpenException seo("Session open failed.");
        seo.causedBy(e);
        throw seo;
    }
    debugLogger->LOG_FINEST("exit");
}

void TLSSession::processGreeting() throw (EPPIOException, GreetingException)
{
    if (greeting.get() == NULL)
    {
        greeting = auto_ptr<Greeting>(new Greeting);
    }

    try
    {
        read(*greeting.get());
    }
    catch (EPPException& pe)
    {
        GreetingException ge("Error while processing greeting.");
        ge.causedBy(pe);
        throw ge;
    }
}

void TLSSession::openSocket()
{
    debugLogger->LOG_FINEST("enter");
    socket = ctx->createSocket(hostName, port, SO_TIMEOUT);
    debugLogger->LOG_FINEST("exit");
}

void TLSSession::login()
    throw (LoginException, EPPIOException, SessionLimitExceededException)
{
    LoginCommand login(username, password, newPW.get(), eppVersion, language,
                       objURIs, extURIs);

    CLTRID::setClID(username);
    writeXML(login.toXML(), login.getCommandType());
    mruTime = Timer::now();
    Response response;

    try
    {
        read(response);
    }
    catch (ParsingException &pe)
    {
        LoginException le("Parse error during login.");
        le.causedBy(pe);
        throw le;
    }

    Result result(response.getResults()[0]);

    string msg = result.getResultMessage();

    switch (result.getResultCode())
    {
    case ResultCode::SUCCESS:
        return;
    case ResultCode::CMD_USE_ERR:
        userLogger->warning(ErrorPkg::getMessage ("epp.login.fail.loggedin"));
        throw LoginException(msg);
    
    case ResultCode::AUTHENT_ERROR_CLOSING:
    case ResultCode::AUTHENT_ERR:
        {
            const string cn = socket->getCertificateCommonName();
            if (cn != username)
            {
                string_vec args;
                args.push_back("<<pw>>");
                args.push_back("<<clID>>");

                string_vec vals;
                vals.push_back(password);
                vals.push_back(username);
                userLogger->severe(
                    ErrorPkg::getMessage("epp.login.fail.auth.pw", args, vals));
                CertificateUserMismatchException certMismatch(username, password);
                LoginException le;
                le.causedBy(certMismatch);
                throw le;
            }
            else
            {
                string_vec args;
                args.push_back("<<clID>>");
                args.push_back("<<cn>>");

                string_vec vals;
                vals.push_back(username);
                vals.push_back(cn);

                userLogger->severe(
                    ErrorPkg::getMessage("epp.login.fail.auth.match",
                                         args, vals));
                UserPassMismatchException userPassMis(msg);
                LoginException le;
                le.causedBy(userPassMis);
                throw le;
            }
        }
        break;
            
    case ResultCode::UNIMPL_OBJ_SVC:
        {
            const XalanNode *node = result.getResultValue();

            if (node)
            {
                msg = c_str(
                    TranscodeToLocalCodePage(dynamic_cast<const XalanText *>(node)->getData()));
            }

            userLogger->severe(
                ErrorPkg::getMessage("epp.login.fail.unimpl.objsvc", "<<uri>>", msg));
            throw LoginException(msg);
        }
        break;
    
    case ResultCode::PARAM_VAL_SYNTAX_ERR:
        throw LoginException(result.getResultMessage());

    case ResultCode::SESS_LIM_EXCEEDED_CLOSING:
        throw SessionLimitExceededException();

    case ResultCode::CMD_FAILED:
    case ResultCode::CMD_FAILED_CLOSING:
        throw LoginException("Login failed because server is closing");
    }
}

void TLSSession::close()
{
    // Closing an already closed Session is a no-op.
    if (!_isOpen) return;
    _isOpen = false;

    try
    {
        logout();
    }
    catch (const LogoutException &le)
    {
        debugLogger->LOG_FINE(le.getMessage());
        debugLogger->LOG_FINE("net.event.socket_closed");
    }
    closeSocket();
}

void TLSSession::changePassword(const string& newPassword)
{
    newPW = auto_ptr<string>(new string(newPassword));
    open();
    close();
}

void TLSSession::logout() throw (LogoutException)
{
    LogoutCommand logout;
    Response response;

    try
    {
        writeXML(logout.toXML(), logout.getCommandType());
        read(response);

        const vector<Result>& results = response.getResults();
        switch (results[0].getResultCode())
        {
        case ResultCode::SUCCESS:
            return;
        case ResultCode::CMD_FAILED:
            throw LogoutException (results[0].getResultMessage());
        }
    }
    catch (const EPPIOException &ioe)
    {
        LogoutException le("IO exception during logout operation.");
        le.causedBy(ioe);
        throw le;
    }
    catch (const ParsingException &pe)
    {
        LogoutException le("Parsing error during logout operation.");
        le.causedBy(pe);
        throw le;
    }
}

void TLSSession::closeSocket()
{
    socket = auto_ptr<TLSSocket>();
}

string TLSSession::read() throw (EPPIOException)
{
    debugLogger->LOG_FINEST ("enter");
    int n = readSize();
    debugLogger->LOG_FINEST ("PDU size: " + StringUtils::makeString(n));
    string data (readData(n));
    supportLogger->LOG_INFO(data);
    debugLogger->LOG_FINEST("exit");
    return data;
}


void TLSSession::read(ReceiveSE& receivedElement) throw (EPPIOException, ParsingException)
{
    auto_ptr<XMLDocument> doc(readToDocument());
    receivedElement.fromXML(doc.get());
}

XMLDocument* TLSSession::readToDocument() 
        throw (EPPIOException, ParsingException)
{
    string xml = read();
    return parser->parse(xml);
}

void TLSSession::writeXML(const string& xml)
    throw (EPPIOException, ParsingException)
{
    if (validationEnabled)
    {
        try
        {
            auto_ptr<XMLDocument> doc(parser->parse(xml));
        }
        catch (ParsingException& e)
        {
            userLogger->warning("Outbound service element malformed: " + xml);
            userLogger->warning(e.getMessage());
            throw e;
        }
    }
    doWrite(xml);
    mruTime = Timer::now();
}

void TLSSession::write(Command& command) throw (EPPIOException, ParsingException)
{
    return writeXML(command.toXML());
}

void TLSSession::doWrite(const string &xml) throw (EPPIOException)
{
    debugLogger->LOG_FINEST("enter");
    try
    {
        if (socket.get() == NULL)
            throw EPPIOException("The socket is not initialised.");
    
        writeSize(xml.length());
        writeData(xml);
        userLogger->info(xml);
    }
    catch (EPPException& e)
    {
        _isInvalid = true;
        EPPIOException eio("An error occured while writing to the socket.");
        eio.causedBy(e);
        throw eio;
    }
    debugLogger->LOG_FINEST("exit");
}

void TLSSession::writeXML(const string& xml, const CommandType* cmdType)
    throw (EPPIOException)
{
    debugLogger->LOG_FINER("writing command " + cmdType->toString());
    doWrite(xml);
    incCommandCounter(cmdType);
}

void TLSSession::keepAlive() throw (EPPIOException)
{
    debugLogger->LOG_FINEST("enter");

    try
    {
        acquire();
        writeXML(pollXML(), poll().getCommandType());
        read();
        release();
    }
    catch (EPPTimeoutException& e)
    {
        userLogger->LOG_WARNING(e.getMessage());
    }
    catch (EPPInterruptedException &ie)
    {
        userLogger->LOG_INFO(ie.getMessage());
    }
    debugLogger->LOG_FINEST("exit");
}

int TLSSession::readSize() throw (EPPIOException)
{
    int size = 0;
    try
    {
        size = socket->readInt() - 4;
    }
    catch(SSLException &e)
    {
        throw EPPIOException("Error when reading data size from socket.");
    }
    return size;
}

void TLSSession::writeSize(int size) throw (EPPIOException)
{
    try
    {
        socket->writeInt(size + 4);
    }
    catch(SSLException &e)
    {
        throw EPPIOException("Error when writing data size to socket.");
    }
}

string TLSSession::readData(int length) throw (EPPIOException)
{
    string inputBuffer;
    try
    {
        socket->readFully(inputBuffer, length);
    }
    catch(SSLException &e)
    {
        throw EPPIOException("Error when reading data from socket.");
    }
    return inputBuffer;
}

void TLSSession::writeData(const string &xml) throw (EPPIOException)
{
    try
    {
        socket->writeBytes(xml);
    }
    catch(SSLException &e)
    {
        throw EPPIOException("Error when writing data to socket.");
    }
}

void TLSSession::incCommandCounter(const CommandType* type)
{
    debugLogger->LOG_FINEST("enter");
    commandCounter->increment(type);
    debugLogger->LOG_FINEST("exit");
}

void TLSSession::incResultCounter(int resultCode)
{
    debugLogger->LOG_FINEST("enter");
    resultCounter.increment(resultCode);
    debugLogger->LOG_FINEST("exit");
}

int TLSSession::getCommandCount()
{
    debugLogger->LOG_FINEST("enter & exit");
    return commandCounter->getTotal();
}

int TLSSession::getCommandCount(const CommandType* type)
{
    debugLogger->LOG_FINEST("enter & exit");
    return commandCounter->getCount(type);
}

int TLSSession::getResultCodeCount(int resultCode)
{
    debugLogger->LOG_FINEST("enter & exit");
    return resultCounter.getValue(resultCode);
}

long TLSSession::getMruInterval() const
{
    return Timer::msDiff(mruTime);
}

void TLSSession::acquire() throw (EPPInterruptedException, EPPTimeoutException)
{
    debugLogger->LOG_FINEST("enter");
    {
        AutoMutex lock(&mtx);
        while (_inUse)
        {
            const struct timespec until(Timer::msOffset2abs(acquireTimeout));

            switch (pthread_cond_timedwait(&cond, &mtx, &until))
            {
                case EINTR:
                    throw EPPInterruptedException(
                            "Interupted while waiting to acquire a session.");
                case ETIMEDOUT:
                    throw EPPTimeoutException(
                            "Timed-out while waiting to acquire session");
                default:
                    // Default implies signalled.
                    break;
            }
            debugLogger->LOG_FINE(ErrorPkg::getMessage("thread.wakeup.info"));
        }
        _inUse = true;
    }
    debugLogger->LOG_FINEST("exit");
}

void TLSSession::release() throw (IllegalStateException)
{
    debugLogger->LOG_FINEST("enter");
    {
        if (!_inUse)
            throw IllegalStateException(
                "Attempt to release a session that was not in use");
        AutoMutex lock(&mtx);
        _inUse = false;
        pthread_cond_signal(&cond);
    }
    debugLogger->LOG_FINEST("exit");
}

File: src/main/java/com/ausregistry/jtoolkit2/session/TLSSession.java
package com.ausregistry.jtoolkit2.session;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.ConnectException;
import java.net.InetAddress;
import java.net.SocketTimeoutException;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

import javax.net.ssl.SSLHandshakeException;
import javax.net.ssl.SSLSocket;

import com.ausregistry.jtoolkit2.xml.XmlOutputConfig;
import org.xml.sax.SAXException;

import com.ausregistry.jtoolkit2.ErrorPkg;
import com.ausregistry.jtoolkit2.Timer;
import com.ausregistry.jtoolkit2.se.CLTRID;
import com.ausregistry.jtoolkit2.se.Command;
import com.ausregistry.jtoolkit2.se.CommandType;
import com.ausregistry.jtoolkit2.se.Greeting;
import com.ausregistry.jtoolkit2.se.LoginCommand;
import com.ausregistry.jtoolkit2.se.LogoutCommand;
import com.ausregistry.jtoolkit2.se.PollRequestCommand;
import com.ausregistry.jtoolkit2.se.Response;
import com.ausregistry.jtoolkit2.se.Result;
import com.ausregistry.jtoolkit2.se.ResultCode;
import com.ausregistry.jtoolkit2.xml.ParsingException;
import com.ausregistry.jtoolkit2.xml.XMLDocument;
import com.ausregistry.jtoolkit2.xml.XMLParser;

/**
 * <p>
 * RFC5734 specifies a transport mapping for EPP over TCP; this class implements that mapping. That specification
 * requires that the session is layered over TLS (Transport Layer Security). This class complies with RFC5734 in
 * implementing the Session interface. It also implements its own StatsManager since it is tightly coupled with the
 * Session implementation.
 * </p>
 *
 * <p>
 * Uses the debug, support and user level loggers.
 * </p>
 */
public class TLSSession implements Session, StatsManager {
    private static final String[] TYPE_INTERVAL_ARR = new String[] {"<<type>>", "<<interval>>" };
    private static final String[] TIME_COUNT_ARR = new String[] {"<<time>>", "<<count>>" };

    private static final int BUF_SIZE = 4096;
    private static String pollXML;
    private static CommandType pollCmdType;

    static {
        PollRequestCommand poll = new PollRequestCommand();
        try {
            pollXML = poll.toXML();
        } catch (SAXException saxe) {
        }
        pollCmdType = poll.getCommandType();
    }

    private java.io.DataInputStream in;
    private java.io.DataOutputStream out;
    private TLSContext ctx;
    private XMLParser parser;
    private SSLSocket socket;

    private boolean inUse;
    private boolean isOpen;
    private boolean isInvalid;
    private CommandCounter commandCounter;
    private ResultCounter resultCounter;
    private long totalTime;
    private Map<CommandType, Long> commandTimeMap;
    private long mruTime;
    private long acquireTimeout;

    private InetAddress inaddr;
    private int port;
    private int soTimeout;
    private String username;
    private String password;
    private String newPW;
    private String eppVersion;
    private String language;
    private String[] objURIs, extURIs;
    private Greeting greeting;

    private String pname;

    private Logger debugLogger;
    private Logger supportLogger;
    private Logger userLogger;
    private boolean needOutputNamespacePrefixInXml;

    {
        pname = TLSSession.class.getPackage().getName();
        debugLogger = Logger.getLogger(pname + ".debug");
        supportLogger = Logger.getLogger(pname + ".support");
        userLogger = Logger.getLogger(pname + ".user");
        parser = new XMLParser();
        resultCounter = new ResultCounter();
        commandTimeMap = new HashMap<CommandType, Long>();
        totalTime = 0L;
        isInvalid = true;
        greeting = null;
    }

    protected TLSSession() {
    }

    protected TLSSession(SessionProperties props) throws SessionConfigurationException {

        configure(props);
    }

    /**
     * Configure the session as described in the Session interface.
     *
     * @throws SessionConfigurationException
     *             Possible causes:
     *             <ul>
     *             <li>KeyStoreNotFoundException</li>
     *             <li>KeyStoreTypeException</li>
     *             <li>NoSuchAlgorithmException</li>
     *             <li>UnrecoverableKeyException</li>
     *             <li>CertificateException</li>
     *             <li>KeyStoreReadException</li>
     *             <li>UnknownHostException</li>
     *             <li>FileNotFoundException</li>
     *             </ul>
     */
    @Override
    public void configure(SessionProperties properties) throws SessionConfigurationException {

        this.port = properties.getPort();
        this.username = properties.getClientID();
        this.password = properties.getClientPW();
        this.newPW = null;
        this.eppVersion = properties.getVersion();
        this.language = properties.getLanguage();
        this.objURIs = properties.getObjURIs();
        this.extURIs = properties.getExtURIs();
        commandCounter = new CommandCounter(properties.getCommandLimitInterval());
        this.acquireTimeout = properties.getAcquireTimeout();
        this.soTimeout = properties.getSocketTimeout();
        this.needOutputNamespacePrefixInXml = properties.needOutputNamespacePrefixInXml();

        try {
            inaddr = InetAddress.getByName(properties.getHostname());

            if (ctx == null) {
                ctx = new TLSContext(properties.getKeyStoreFilename(), properties.getKeyStorePassphrase(),
                        properties.getTrustStoreFilename(), properties.getTrustStorePassphrase(),
                        properties.getKeyStoreType(), properties.getSSLAlgorithm(), properties.getSSLVersion());
            }
        } catch (Exception e) {
            throw new SessionConfigurationException(e);
        }
    }

    @Override
    public boolean isInvalid() {
        return isInvalid;
    }

    @Override
    public boolean isOpen() {
        return isOpen;
    }

    /**
     * * An EPP session is opened by first establishing a connection using the server location information and
     * authentication sources provided in <code>configure</code>, then issuing a login command with further
     * authentication data and options provided in <code>configure</code>. Service information provided immediately upon
     * connection establishment may affect options provided in the login command.
     *
     * @throws SessionOpenException
     *             The getCause() method should be invoked on the exception thrown. The cause may be one of:
     *             <dl>
     *             <dt>SSLHandshakeException</dt>
     *             <dd>The SSL handshake failed. The reason is described in the exception message, and is also recorded
     *             in the user logs.</dd>
     *             <dt>IOException</dt>
     *             <dd>See the log record published to the user logs, or check the exception message.</dd>
     *             <dt>GreetingException</dt>
     *             <dd>The service element received from the server upon connection establishment was not a valid EPP
     *             greeting.</dd>
     *             <dt>LoginException</dt>
     *             <dd>The login command failed to establish an EPP session. The cause, available via getCause(),
     *             describes the specific reason for failure.</dd>
     *             </dl>
     */
    @Override
    public void open() throws SessionOpenException {
        isInvalid = true;
        try {
            openSocket();
            processGreeting();
            login();
            isOpen = true;
            isInvalid = false;
        } catch (SSLHandshakeException e) {
            userLogger.severe(ErrorPkg.getMessage("TLSContext.createSocket.0", "<<java.home>>",
                    System.getProperty("java.home", "java.home")));
            userLogger.severe(e.getMessage());
            throw new SessionOpenException(e);
        } catch (ConnectException e) {
            final String errorMessage = ErrorPkg.getMessage("net.socket.open.fail", new String[] {"<<port>>",
                    "<<host>>" }, new String[] {String.valueOf(port), inaddr.getHostAddress() });
            userLogger.severe(errorMessage);
            userLogger.severe(e.getMessage());
            throw new SessionOpenException(errorMessage);
        } catch (IOException e) {
            userLogger.severe(ErrorPkg.getMessage("net.socket.open.fail", new String[] {"<<port>>", "<<host>>" },
                    new String[] {String.valueOf(port), inaddr.getHostAddress() }));
            userLogger.severe(e.getMessage());
            throw new SessionOpenException(ErrorPkg.getMessage("net.socket.open.fail", new String[] {"<<port>>",
                    "<<host>>" }, new String[] {String.valueOf(port), inaddr.getHostAddress() }));
        } catch (SessionLimitExceededException e) {
            e.printStackTrace();
            throw new SessionOpenException(e);
        } catch (GreetingException e) {
            e.printStackTrace();
            throw new SessionOpenException(e);
        } catch (LoginException e) {
            e.printStackTrace();
            throw new SessionOpenException(e);
        }
    }

    @Override
    public void changePassword(String newPassword) throws SessionOpenException {
        newPW = newPassword;
        open();
        close();
    }

    private void openSocket() throws SSLHandshakeException, IOException {
        socket = ctx.createSocket(inaddr.getHostAddress(), port, soTimeout);
        in = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
        out = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream(), BUF_SIZE));
    }

    @Override
    public Greeting getGreeting() {
        return greeting;
    }

    private void processGreeting() throws SessionLimitExceededException, GreetingException {

        try {
            if (greeting == null) {
                greeting = new Greeting();
            }

            greeting.fromXML(readToDocument());
        } catch (IOException pe) {
            throw new SessionLimitExceededException();
        } catch (ParsingException pe) {
            pe.printStackTrace();
            throw new GreetingException(pe);
        }
    }

    private void login() throws LoginException, IOException {
        LoginCommand login = new LoginCommand(username, password, newPW, eppVersion, language, objURIs, extURIs);
        CLTRID.setClID(username);
        long startTime = Timer.now();
        try {
            write(login.toXML(), login.getCommandType());
        } catch (SAXException saxe) {
            throw new LoginException(saxe);
        }
        mruTime = Timer.now();
        Response response = new Response();
        try {
            response.fromXML(readToDocument());
            long responseTime = Timer.msDiff(startTime);
            recordResponseTime(login.getCommandType(), responseTime);
        } catch (NumberFormatException nfe) {
            // These two exceptions usually indicate a fatal error.
            throw new LoginException(nfe);
        } catch (ParsingException pe) {
            throw new LoginException(pe);
        }

        Result result = response.getResults()[0];
        assert result != null;
        String msg = result.getResultMessage();
        switch (result.getResultCode()) {
        case ResultCode.SUCCESS:
            return;
        case ResultCode.CMD_USE_ERR:
            userLogger.warning(ErrorPkg.getMessage("epp.login.fail.loggedin"));
            throw new LoginException(msg);
        case ResultCode.AUTHENT_ERROR_CLOSING:
        case ResultCode.AUTHENT_ERR:
            String cn = ctx.getCertificateCommonName();

            if (cn.equals(username)) {
                userLogger.severe(ErrorPkg.getMessage("epp.login.fail.auth.pw",
                        new String[] {"<<clID>>", "<<pw>>" }, new String[] {username, password }));
                throw new UserPassMismatchException(msg);
            } else {
                userLogger.severe(ErrorPkg.getMessage("epp.login.fail.auth.match", new String[] {"<<clID>>",
                        "<<cn>>" }, new String[] {username, cn }));
                throw new CertificateUserMismatchException(username, cn);
            }
        case ResultCode.UNIMPL_OBJ_SVC:
            if (result.getResultValue() != null) {
                msg = result.getResultValue().item(0).getTextContent();
            }

            userLogger.severe(ErrorPkg.getMessage("epp.login.fail.unimpl.objsvc", "<<uri>>", msg));
            throw new LoginException(msg);
        case ResultCode.PARAM_VAL_SYNTAX_ERR:
            throw new LoginException(new ParameterSyntaxException(result.getResultMessage()));
        case ResultCode.SESS_LIM_EXCEEDED_CLOSING:
            throw new SessionLimitExceededException();
        case ResultCode.CMD_FAILED:
            raiseLoginException(result.hasResultExtReasons(), result.getResultExtReason(0));
        case ResultCode.CMD_FAILED_CLOSING:
            raiseLoginException(result.hasResultExtReasons(), result.getResultExtReason(0));
        default: // do nothing
        }
    }

    private void raiseLoginException(boolean hasValues, String msg) throws LoginException {

        if (hasValues) {
            throw new LoginException(new CommandFailedException(msg));
        } else {
            throw new LoginException(new CommandFailedException());
        }
    }

    @Override
    public void close() {
        inUse = true;
        isOpen = false;
        try {
            logout();
        } catch (LogoutException le) {
            debugLogger.info(le.getMessage());
            debugLogger.info(ErrorPkg.getMessage("net.event.socket_closed"));
        }
        closeSocket();
        inUse = false;
    }

    private void logout() throws LogoutException {
        LogoutCommand logout = new LogoutCommand();
        Response response = new Response();

        try {
            long startTime = Timer.now();

            try {
                write(logout.toXML(), logout.getCommandType());
            } catch (SAXException saxe) {
                debugLogger.warning(saxe.getMessage());
                throw new LogoutException(saxe);
            }
            response.fromXML(readToDocument());
            long responseTime = Timer.msDiff(startTime);
            recordResponseTime(logout.getCommandType(), responseTime);

            Result[] results = response.getResults();
            if (ResultCode.CMD_FAILED == results[0].getResultCode()) {
                throw new LogoutException(new CommandFailedException(results[0].getResultMessage()));
            }
        } catch (IOException ioe) {
            throw new LogoutException(ioe);
        } catch (ParsingException pe) {
            throw new LogoutException(pe);
        }
    }

    private void closeSocket() {
        try {
            in.close();
        } catch (IOException ioe) {
            userLogger.warning(ioe.getMessage());
        }

        try {
            out.close();
        } catch (IOException ioe) {
            userLogger.warning(ioe.getMessage());
        }

        try {
            socket.close();
        } catch (IOException ioe) {
            userLogger.warning(ioe.getMessage());
        }
    }

    /**
     * Receive data from the peer. This method is unsynchronised; the caller MUST provide synchronisation against other
     * calls to read.
     *
     * @return the details from the socket
     * @throws IOException Signals that an I/O exception has occurred.
     */
    @Override
    public String read() throws IOException {
        try {
            int n = readSize();
            debugLogger.finer("PDU size: " + n);
            String data = readData(n);
            supportLogger.info(data);
            return data;
        } catch (SocketTimeoutException ste) {
            userLogger.warning(ste.getMessage());
            userLogger.warning(ErrorPkg.getMessage("epp.session.read.timeout"));
            throw ste;
        } catch (IOException ioe) {
            userLogger.warning(ioe.getMessage());
            throw ioe;
        }
    }

    @Override
    public void read(Response response) throws IOException, ParsingException {
        response.fromXML(readToDocument());
    }

    @Override
    public XMLDocument readToDocument() throws IOException, ParsingException {
        String xml = read();
        assert parser != null;
        return parser.parse(xml);
    }

    /**
     * Send data to peer. This method is unsynchronised; the caller MUST provide synchronisation against other calls to
     * <code>write(String)</code>.
     *
     * @param xml
     *            the XML to be sent to the EPP Server
     * @throws IOException
     *             Signals that an I/O exception has occurred.
     */
    @Override
    public void write(String xml) throws IOException {
        doWrite(xml);
        mruTime = Timer.now();
    }

    @Override
    public void write(Command command) throws IOException, ParsingException {
        try {
            if (needOutputNamespacePrefixInXml) {
                write(command.toXML(XmlOutputConfig.prefixAllNamespaceConfig()));
            } else {
                write(command.toXML());
            }
        } catch (SAXException saxe) {
            throw new ParsingException(saxe);
        }
    }

    private void doWrite(String xml) throws IOException {
        if (out == null) {
            throw new UninitialisedSessionException();
        }

        try {
            final byte[] xmlBytes = xml.getBytes();
            writeSize(xmlBytes.length);
            writeData(xmlBytes);
        } catch (IOException ioe) {
            isInvalid = true;
            throw ioe;
        }
    }

    private void write(String xml, CommandType cmdType) throws IOException {
        debugLogger.finer("writing command " + cmdType.toString());
        doWrite(xml);
        incCommandCounter(cmdType);
    }

    /**
     * Send a poll command to the EPP server in order to prevent the session timing out. This operation does not affect
     * the most-recently-used statistic.
     */
    @Override
    public void keepAlive() throws IOException {
        try {
            acquire();
            write(pollXML, pollCmdType);
            read(); // not interested in response.
            release();
        } catch (TimeoutException te) {
            userLogger.info(te.getMessage());
        } catch (InterruptedException ie) {
            userLogger.info(ie.getMessage());
        }
    }

    private int readSize() throws IOException {
        return in.readInt() - 4;
    }

    private void writeSize(int size) throws IOException {
        out.writeInt(size + 4);
    }

    private String readData(int length) throws IOException {
        byte[] inputBuffer = new byte[length];
        in.readFully(inputBuffer, 0, length);

        return new String(inputBuffer);
    }

    private void writeData(final byte[] xml) throws IOException {
        out.write(xml);
        out.flush();
    }

    @Override
    public void incCommandCounter(CommandType type) {
        commandCounter.increment(type);
    }

    @Override
    public void incResultCounter(int resultCode) {
        resultCounter.increment(resultCode);
    }

    @Override
    public void recordResponseTime(CommandType type, long responseTime) {
        totalTime += responseTime;
        if (!commandTimeMap.containsKey(type)) {
            commandTimeMap.put(type, responseTime);
            debugLogger.info(ErrorPkg.getMessage("epp.server.response_time.new_cmd", TYPE_INTERVAL_ARR, new String[] {
                    type.getCommandName(), String.valueOf(responseTime) }));
        } else {
            commandTimeMap.put(type, commandTimeMap.get(type) + responseTime);
            debugLogger.info(ErrorPkg.getMessage("epp.server.response_time.previous_cmd", TYPE_INTERVAL_ARR,
                    new String[] {type.getCommandName(), String.valueOf(responseTime) }));
        }
    }

    @Override
    public long getAverageResponseTime() {
        long totalCount = commandCounter.getTotal();
        if (totalCount == 0L) {
            return 0L;
        }
        debugLogger.info(ErrorPkg.getMessage("epp.server.response_time.avg", TIME_COUNT_ARR,
                new String[] {String.valueOf(totalTime), String.valueOf(totalCount) }));

        return totalTime / totalCount;
    }

    @Override
    public long getAverageResponseTime(CommandType type) {
        if (!commandTimeMap.containsKey(type)) {
            return 0L;
        }

        long cmdCount = commandCounter.getCount(type);

        if (cmdCount == 0L) {
            return 0L;
        }

        return commandTimeMap.get(type) / cmdCount;
    }

    @Override
    public long getCommandCount() {
        return commandCounter.getTotal();
    }

    @Override
    public long getCommandCount(CommandType type) {
        return commandCounter.getCount(type);
    }

    @Override
    public int getRecentCommandCount() {
        return commandCounter.getExactRecentTotal();
    }

    @Override
    public int getRecentCommandCount(CommandType type) {
        return commandCounter.getRecentCount(type);
    }

    @Override
    public long getResultCodeCount(int resultCode) {
        return resultCounter.getValue(resultCode);
    }

    /**
     * Get the length of time (in milliseconds) since the most recent use (mru) of the session. The session is
     * considered to be used when the write method is invoked.
     */
    @Override
    public long getMruInterval() {
        return Timer.msDiff(mruTime);
    }

    @Override
    public StatsManager getStatsManager() {
        return this;
    }

    @Override
    public boolean isAvailable() {
        return (isOpen && !inUse);
    }

    @Override
    public void acquire() throws InterruptedException, TimeoutException {
        synchronized (this) {
            while (inUse) {
                wait(acquireTimeout);
                if (inUse) {
                    throw new TimeoutException(ErrorPkg.getMessage("epp.session.acquire.timeout", "<<timeout>>",
                            String.valueOf(acquireTimeout)));
                }
            }
            inUse = true;
        }
    }

    @Override
    public void release() {
        synchronized (this) {
            inUse = false;
            notify();
        }
    }
}
